### DISCUSSION TITLE
FR: Make asset custom fields unique to an entry

### DISCUSSION THREAD

**Discussion by richhayler** [craftcms/cms#8509] [Ideas]
### Description
Assets can have their own custom fields - for example you can create a plain text field to hold the image caption. In this case the text entered into the image caption would be accessible in any entry the image was referenced from. It would be great if the data included in the asset’s custom fields could be unique on an individual entry basis. A specific use case is having custom asset fields for hero images that could specify text overlay position, alignment, theme, etc. While these fields could be added to the entry form itself it would make sense to associate them with the image as in the abscence of an image they may not be relevant. Allowing the data contained within these custom asset fields to be unique to each entry would mean you could use the same image across multiple entries with different values applied.

**Comment by brandonkelly** [comment]
In lieu of this, you can create a Matrix field with one block type, with an Assets field (limited to one selection) and whatever other custom fields you need to be associated with each asset. Then each block ends up representing a relation, but also contains the entry-specific relation data you’re after.

**Comment by richhayler** [comment]
That’s a reasonable approach. Personally I think the asset custom fields are a really useful and powerful addition to craft’s flexibility but in its current form it can mean that content authors can inadvertently reference an image in an entry (that has already been used for a different entry) and update its custom fields without realising the implications.

**Comment by brandonkelly** [comment]
Agree your idea could provide a better AX.

**Comment by hiasl** [comment]
My suggestion for a universal solution:
- Allow fields to be assigned to relations (to entries, assets, users, ...)
- This should happen in the field definition process, which creates the relation in the entry. This means when you create e.g. an asset field, you should be able to assign other fields to it for the relation it creates
- These fields are then available to be filled out for each asset assigned when you edit the entry that contains the asset field.
- Fields to be assignable to relations should be the same as to entry types
- When an element to which a relation is created, contains the same field, its value would be overridden (maybe also cancelable)
- When an element to which a relation is created does not contain the field from the relation, its value is available in addition
- The GUI should stay the same as it currently is with all relations. Double clicking the relation slides in the modal from the right side. Fields from the relation are edited inside of the modal in addition to the fields from the related element.
